### ISSUE TITLE
Adding/removing Okta user to a OIN Okta assignment or push group, unintended add/removes take place

### ISSUE THREAD

**Issue by laiandrewe** [okta/terraform-provider-okta#382]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v0.13.5
+ provider registry.terraform.io/hashicorp/null v3.0.0
+ provider registry.terraform.io/oktadeveloper/okta v3.10.1

### Affected Resource(s)

- okta_user
- okta_group

### Terraform Configuration Files

```hcl
resource "okta_user" "example_user" {
  admin_roles = [
    "SUPER_ADMIN",
  ]
  custom_profile_attributes = jsonencode({})
  display_name              = "Example User"
  email                     = "exampleuser@example.com"
  first_name                = "Example"
  last_name                 = "User"
  login                     = "exampleuser"
  group_memberships = [
    okta_group.example1.id,
    okta_group.example2.id,
    okta_group.example3.id,
-   okta_group.example3.id,
}
```

### Expected Behavior

We understand there's no API for OIN apps and push group support is not available, so we treat as a standard Okta group and manage in Terraform.
1. Terraform plan shows user is added/removed from the intended group
2. User is added/removed from Okta group (Assignment or Push group within Okta OIN app).
2. Group is pushed to external application

### Actual Behavior

After terraform apply, Okta logs show:

1. Terraform plan shows user is added/removed from group
2. After Terraform apply:
2. Remove user’s application membership
2. Group Push - multiple push group calls made
3. Push user deactivation to external application
4. Group Push - multiple push group calls made
5. success
6. Group Push - multiple push group calls made
7. success
8. Group Push
9. Add user to application membership
10. Group Push - multiple push group calls made
11. Verify user exists in external application
12. Push user reactivation in external application
13. Group Push - multiple push group calls made

In other OIN apps the user is assigned:

1. Remove user's application membership
2. Add user to application membership
3. Verify user exists in external application

### Steps to Reproduce

1. add/remove user from group - This group is a push group for a OIN Okta app
2. terraform plan shows expected change of add/remove user from specified group
3. terraform apply

### Important Factoids

When adding/removing a user from a push/assignment group assigned to a Okta OIN application, this action removes and re-adds that user from all OIN apps that the user is assigned to.

**Comment by bogdanprodan-okta** [comment]
Hi, @laiandrewe! Thanks for submitting this issue! I'll try to reproduce the flow to see what can be fixed and get back to you asap.

**Comment by laiandrewe** [comment]
Hello, @bogdanprodan-okta. Wanted to follow up to see if you were able to reproduce this issue?  Thanks!

**Comment by laiandrewe** [comment]
@bogdanprodan-okta Have you been able to reproduce this issue?  I'm surprised other orgs haven't run into this if they're using both Terraform and the Okta OIN applications

**Comment by bogdanprodan-okta** [comment]
Hi, @laiandrewe! Sorry for the late response! I was distracted from terraform provider for a while. I will start to work on this issue tomorrow and see what I can do.

**Comment by bogdanprodan-okta** [comment]
Hi, @laiandrewe! Could you test adding/removing user to/from the group by using this resource?

```
resource "okta_user" "test" {
  first_name = "TestAcc"
  last_name  = "Smith"
  login      = "testAcc-replace_with_uuid@example.com"
  email      = "testAcc-replace_with_uuid@example.com"

  lifecycle {
    ignore_changes = [group_memberships] // so the terraform plan will skip this
  }
}

resource "okta_user_group_memberships" "test" {
  user_id = okta_user.test.id
  groups = [
    okta_group.test_1.id,
    okta_group.test_2.id,
  ]
}
```
I just want to make sure that the behavior is the same

**Comment by bogdanprodan-okta** [comment]
I think I found the root cause of this issue. Here is the code sample responsible for that:
```
// need to remove from all current groups and reassign based on terraform configs when a change is detected
func updateGroupsOnUser(ctx context.Context, u string, g []string, c *okta.Client) error {
	groups, _, err := c.User.ListUserGroups(ctx, u)
	if err != nil {
		return fmt.Errorf("failed to list user groups: %v", err)
	}
	for _, group := range groups {
		if group.Type != "BUILT_IN" && group.Type != "APP_GROUP" {
			_, err = c.Group.RemoveUserFromGroup(ctx, group.Id, u)
			if err != nil {
				return fmt.Errorf("failed to remove user from group: %v", err)
			}
		}
	}
	return assignGroupsToUser(ctx, u, g, c)
}
```
That's why you see all these unnecessary assigns/unassigns.

I would suggest managing group assignments via `okta_user_group_memberships` resource because it has different behavior than the `group_memberships` field in the `okta_user` resource.

If you have any other questions, please feel free to post them here. Thanks! And thanks for the patience!

P.S.`group_memberships` field will be removed in the 4.x version of the terraform and all the app/group memberships will be managed via separate resources.

**Comment by laiandrewe** [comment]
Hi @bogdanprodan-okta thanks for the update.  I'm attempting to test this but I receive `Error: Invalid resource type` for `resource "okta_user_group_memberships" "test"`

**Comment by bogdanprodan-okta** [comment]
> Hi @bogdanprodan-okta thanks for the update. I'm attempting to test this but I receive `Error: Invalid resource type` for `resource "okta_user_group_memberships" "test"`

Could you use the `v3.11.0` provider from GitHub? The problem is that the latest published version is `v3.10.1` and we can not yet push the newest to the registry. See the explanation [here](https://github.com/okta/terraform-provider-okta#important-note).
